File: backend/app/schemas/subscriber.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.coin_transaction import TransactionType
from app.models.subscriber import SubscriberStatus


class SubscriberUpdate(BaseModel):
    status: SubscriberStatus | None = None
    admin_notes: str | None = None


class CoinAdjustmentRequest(BaseModel):
    amount: int = Field(..., description="Positive to credit, negative to debit")
    description: str = Field(..., min_length=1, max_length=500)


class SubscriberResponse(BaseModel):
    id: uuid.UUID
    email: str | None
    name: str | None
    country: str | None
    avatar_url: str | None
    device_id: str
    status: SubscriberStatus
    coin_balance: int
    registered_at: datetime | None
    last_active_at: datetime | None
    admin_notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class SubscriberListResponse(BaseModel):
    items: list[SubscriberResponse]
    total: int
    page: int
    per_page: int


class CoinTransactionResponse(BaseModel):
    id: uuid.UUID
    subscriber_id: uuid.UUID
    type: TransactionType
    amount: int
    balance_after: int
    reference_type: str | None
    reference_id: uuid.UUID | None
    description: str | None
    created_by: uuid.UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


class CoinTransactionListResponse(BaseModel):
    items: list[CoinTransactionResponse]
    total: int
    page: int
    per_page: int

File: frontend/src/types/subscriber.ts
export type SubscriberStatus = 'anonymous' | 'active' | 'suspended' | 'banned'
export type TransactionType = 'purchase' | 'spend' | 'refund' | 'promo' | 'adjustment'

export interface Subscriber {
  id: string
  email: string | null
  name: string | null
  country: string | null
  avatar_url: string | null
  device_id: string
  status: SubscriberStatus
  coin_balance: number
  registered_at: string | null
  last_active_at: string | null
  admin_notes: string | null
  created_at: string
  updated_at: string
}

export interface SubscriberListResponse {
  items: Subscriber[]
  total: number
  page: number
  per_page: number
}

export interface SubscriberUpdate {
  status?: SubscriberStatus
  admin_notes?: string
}

export interface CoinAdjustment {
  amount: number
  description: string
}

export interface CoinTransaction {
  id: string
  subscriber_id: string
  type: TransactionType
  amount: number
  balance_after: number
  reference_type: string | null
  reference_id: string | null
  description: string | null
  created_by: string | null
  created_at: string
}

export interface CoinTransactionListResponse {
  items: CoinTransaction[]
  total: number
  page: number
  per_page: number
}

export interface CoinPackage {
  id: string
  name: string
  description: string | null
  coin_amount: number
  price_sar: string
  is_active: boolean
  sort_order: number
  created_by: string
  created_at: string
  updated_at: string
}

export interface CoinPackageListResponse {
  items: CoinPackage[]
  total: number
}

export interface CoinPackageCreate {
  name: string
  description?: string
  coin_amount: number
  price_sar: string
  sort_order?: number
}

export interface CoinPackageUpdate {
  name?: string
  description?: string
  coin_amount?: number
  price_sar?: string
  is_active?: boolean
  sort_order?: number
}
